File: src/wasm/logic.c
/**
 * NumJS-WASM Logic & Comparison Functions Implementation
 *
 * Provides truth testing, element-wise predicates, and array comparison operations.
 *
 * Reference implementations:
 * - numpy/_core/fromnumeric.py (all, any)
 * - numpy/_core/numeric.py (isclose, allclose, array_equal, array_equiv)
 * - numpy/lib/_ufunclike_impl.py (isneginf, isposinf)
 */

#include "logic.h"
#include "broadcast.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#define EXPORT EMSCRIPTEN_KEEPALIVE
#else
#define EXPORT
#endif

/* ============ Element-wise Predicates ============ */

EXPORT NDArray* ndarray_isfinite(NDArray* arr)
{
    if (!arr) return NULL;

    /* Create bool result with same shape */
    NDArray* result = ndarray_empty(arr->ndim, arr->shape, DTYPE_BOOL);
    if (!result) return NULL;

    uint8_t* out = (uint8_t*)result->data;

    for (size_t i = 0; i < arr->size; i++) {
        double val = ndarray_get_flat(arr, i);
        out[i] = isfinite(val) ? 1 : 0;
    }

    return result;
}

EXPORT NDArray* ndarray_isinf(NDArray* arr)
{
    if (!arr) return NULL;

    NDArray* result = ndarray_empty(arr->ndim, arr->shape, DTYPE_BOOL);
    if (!result) return NULL;

    uint8_t* out = (uint8_t*)result->data;

    for (size_t i = 0; i < arr->size; i++) {
        double val = ndarray_get_flat(arr, i);
        out[i] = isinf(val) ? 1 : 0;
    }

    return result;
}

EXPORT NDArray* ndarray_isnan(NDArray* arr)
{
    if (!arr) return NULL;

    NDArray* result = ndarray_empty(arr->ndim, arr->shape, DTYPE_BOOL);
    if (!result) return NULL;

    uint8_t* out = (uint8_t*)result->data;

    for (size_t i = 0; i < arr->size; i++) {
        double val = ndarray_get_flat(arr, i);
        out[i] = isnan(val) ? 1 : 0;
    }

    return result;
}

EXPORT NDArray* ndarray_isneginf(NDArray* arr)
{
    if (!arr) return NULL;

    NDArray* result = ndarray_empty(arr->ndim, arr->shape, DTYPE_BOOL);
    if (!result) return NULL;

    uint8_t* out = (uint8_t*)result->data;

    for (size_t i = 0; i < arr->size; i++) {
        double val = ndarray_get_flat(arr, i);
        out[i] = (isinf(val) && val < 0) ? 1 : 0;
    }

    return result;
}

EXPORT NDArray* ndarray_isposinf(NDArray* arr)
{
    if (!arr) return NULL;

    NDArray* result = ndarray_empty(arr->ndim, arr->shape, DTYPE_BOOL);
    if (!result) return NULL;

    uint8_t* out = (uint8_t*)result->data;

    for (size_t i = 0; i < arr->size; i++) {
        double val = ndarray_get_flat(arr, i);
        out[i] = (isinf(val) && val > 0) ? 1 : 0;
    }

    return result;
}

EXPORT NDArray* ndarray_iscomplex_elem(NDArray* arr)
{
    if (!arr) return NULL;

    NDArray* result = ndarray_empty(arr->ndim, arr->shape, DTYPE_BOOL);
    if (!result) return NULL;

    uint8_t* out = (uint8_t*)result->data;

    /* Non-complex types: all False */
    if (arr->dtype != DTYPE_COMPLEX64 && arr->dtype != DTYPE_COMPLEX128) {
        memset(out, 0, result->size);
        return result;
    }

    /* Complex: check imag != 0 */
    for (size_t i = 0; i < arr->size; i++) {
        double imag = ndarray_get_complex_imag(arr, i);
        out[i] = (imag != 0.0) ? 1 : 0;
    }

    return result;
}

EXPORT NDArray* ndarray_isreal_elem(NDArray* arr)
{
    if (!arr) return NULL;

    NDArray* result = ndarray_empty(arr->ndim, arr->shape, DTYPE_BOOL);
    if (!result) return NULL;

    uint8_t* out = (uint8_t*)result->data;

    /* Non-complex types: all True */
    if (arr->dtype != DTYPE_COMPLEX64 && arr->dtype != DTYPE_COMPLEX128) {
        memset(out, 1, result->size);
        return result;
    }

    /* Complex: check imag == 0 */
    for (size_t i = 0; i < arr->size; i++) {
        double imag = ndarray_get_complex_imag(arr, i);
        out[i] = (imag == 0.0) ? 1 : 0;
    }

    return result;
}

/* ============ Reductions: all / any ============ */

EXPORT NDArray* ndarray_all(NDArray* arr)
{
    if (!arr) return NULL;

    /* Create 0-d bool result, initialized to False */
    NDArray* result = ndarray_scalar(0.0, DTYPE_BOOL);
    if (!result) return NULL;

    /* Check all elements */
    for (size_t i = 0; i < arr->size; i++) {
        double val = ndarray_get_flat(arr, i);
        if (val == 0.0) {
            /* Found a falsy value - return False */
            return result;
        }
    }

    /* All truthy - return True */
    ndarray_set_flat(result, 0, 1.0);
    return result;
}

EXPORT NDArray* ndarray_all_axis(NDArray* arr, int32_t axis, int keepdims)
{
    if (!arr) return NULL;

    /* Normalize negative axis */
    if (axis < 0) axis += arr->ndim;
    if (axis < 0 || axis >= arr->ndim) return NULL;

    /* Compute result shape */
    int32_t result_shape[NPY_MAXDIMS];
    int32_t result_ndim;

    if (keepdims) {
        result_ndim = arr->ndim;
        for (int i = 0; i < arr->ndim; i++) {
            result_shape[i] = (i == axis) ? 1 : arr->shape[i];
        }
    } else {
        result_ndim = arr->ndim - 1;
        int j = 0;
        for (int i = 0; i < arr->ndim; i++) {
            if (i != axis) {
                result_shape[j++] = arr->shape[i];
            }
        }
    }

    /* Handle 0-d result (when arr is 1-d) */
    if (result_ndim == 0) {
        return ndarray_all(arr);
    }

    /* Initialize result to True (1) */
    NDArray* result = ndarray_full(result_ndim, result_shape, DTYPE_BOOL, 1.0);
    if (!result) return NULL;

    /* Iterate and reduce along axis */
    int32_t indices[NPY_MAXDIMS] = {0};
    int32_t result_indices[NPY_MAXDIMS];

    for (size_t flat = 0; flat < arr->size; flat++) {
        /* Map arr indices to result indices (skipping axis) */
        int j = 0;
        for (int i = 0; i < arr->ndim; i++) {
            if (keepdims) {
                result_indices[i] = (i == axis) ? 0 : indices[i];
            } else if (i != axis) {
                result_indices[j++] = indices[i];
            }
        }

        /* Get current value and result value */
        double val = ndarray_get_item(arr, indices, arr->ndim);
        double cur = ndarray_get_item(result, result_indices, result_ndim);

        /* Logical AND: if val is 0, result becomes 0 */
        if (cur != 0.0 && val == 0.0) {
            ndarray_set_item(result, result_indices, result_ndim, 0.0);
        }

        /* Increment indices */
        for (int d = arr->ndim - 1; d >= 0; d--) {
            indices[d]++;
            if (indices[d] < arr->shape[d]) break;
            indices[d] = 0;
        }
    }

    return result;
}

EXPORT NDArray* ndarray_any(NDArray* arr)
{
    if (!arr) return NULL;

    /* Create 0-d bool result, initialized to False */
    NDArray* result = ndarray_scalar(0.0, DTYPE_BOOL);
    if (!result) return NULL;

    /* Check all elements */
    for (size_t i = 0; i < arr->size; i++) {
        double val = ndarray_get_flat(arr, i);
        if (val != 0.0) {
            /* Found a truthy value - return True */
            ndarray_set_flat(result, 0, 1.0);
            return result;
        }
    }

    /* All falsy - return False */
    return result;
}

EXPORT NDArray* ndarray_any_axis(NDArray* arr, int32_t axis, int keepdims)
{
    if (!arr) return NULL;

    /* Normalize negative axis */
    if (axis < 0) axis += arr->ndim;
    if (axis < 0 || axis >= arr->ndim) return NULL;

    /* Compute result shape */
    int32_t result_shape[NPY_MAXDIMS];
    int32_t result_ndim;

    if (keepdims) {
        result_ndim = arr->ndim;
        for (int i = 0; i < arr->ndim; i++) {
            result_shape[i] = (i == axis) ? 1 : arr->shape[i];
        }
    } else {
        result_ndim = arr->ndim - 1;
        int j = 0;
        for (int i = 0; i < arr->ndim; i++) {
            if (i != axis) {
                result_shape[j++] = arr->shape[i];
            }
        }
    }

    /* Handle 0-d result (when arr is 1-d) */
    if (result_ndim == 0) {
        return ndarray_any(arr);
    }

    /* Initialize result to False (0) */
    NDArray* result = ndarray_create(result_ndim, result_shape, DTYPE_BOOL);
    if (!result) return NULL;

    /* Iterate and reduce along axis */
    int32_t indices[NPY_MAXDIMS] = {0};
    int32_t result_indices[NPY_MAXDIMS];

    for (size_t flat = 0; flat < arr->size; flat++) {
        /* Map arr indices to result indices (skipping axis) */
        int j = 0;
        for (int i = 0; i < arr->ndim; i++) {
            if (keepdims) {
                result_indices[i] = (i == axis) ? 0 : indices[i];
            } else if (i != axis) {
                result_indices[j++] = indices[i];
            }
        }

        /* Get current value and result value */
        double val = ndarray_get_item(arr, indices, arr->ndim);
        double cur = ndarray_get_item(result, result_indices, result_ndim);

        /* Logical OR: if val is non-zero, result becomes 1 */
        if (cur == 0.0 && val != 0.0) {
            ndarray_set_item(result, result_indices, result_ndim, 1.0);
        }

        /* Increment indices */
        for (int d = arr->ndim - 1; d >= 0; d--) {
            indices[d]++;
            if (indices[d] < arr->shape[d]) break;
            indices[d] = 0;
        }
    }

    return result;
}

/* ============ Comparison Functions ============ */

EXPORT NDArray* ndarray_isclose(NDArray* a, NDArray* b, double rtol, double atol, int equal_nan)
{
    if (!a || !b) return NULL;

    /* Compute broadcast shape */
    int32_t bc_shape[NPY_MAXDIMS];
    int32_t bc_ndim;

    if (broadcast_shapes(a->shape, a->ndim, b->shape, b->ndim, bc_shape, &bc_ndim) != 0) {
        return NULL;
    }

    /* Create result array */
    NDArray* result = ndarray_empty(bc_ndim, bc_shape, DTYPE_BOOL);
    if (!result) return NULL;

    /* Compute broadcast strides */
    int32_t a_strides[NPY_MAXDIMS];
    int32_t b_strides[NPY_MAXDIMS];

    if (broadcast_strides(a, bc_shape, bc_ndim, a_strides) != 0 ||
        broadcast_strides(b, bc_shape, bc_ndim, b_strides) != 0) {
        ndarray_free(result);
        return NULL;
    }

    size_t a_elem_size = dtype_size(a->dtype);
    size_t b_elem_size = dtype_size(b->dtype);

    uint8_t* out = (uint8_t*)result->data;
    int32_t indices[NPY_MAXDIMS] = {0};

    for (size_t i = 0; i < result->size; i++) {
        /* Compute flat offsets using broadcast strides */
        size_t a_offset = 0;
        size_t b_offset = 0;

        for (int d = 0; d < bc_ndim; d++) {
            if (a_strides[d] != 0) {
                a_offset += (size_t)indices[d] * ((size_t)a_strides[d] / a_elem_size);
            }
            if (b_strides[d] != 0) {
                b_offset += (size_t)indices[d] * ((size_t)b_strides[d] / b_elem_size);
            }
        }

        double val_a = ndarray_get_flat(a, a_offset);
        double val_b = ndarray_get_flat(b, b_offset);

        /* Check for NaN equality */
        int a_nan = isnan(val_a);
        int b_nan = isnan(val_b);

        if (equal_nan && a_nan && b_nan) {
            out[i] = 1;
        } else if (a_nan || b_nan) {
            out[i] = 0;
        } else if (val_a == val_b) {
            /* Exact equality (handles inf == inf) */
            out[i] = 1;
        } else if (isfinite(val_b)) {
            /* Tolerance comparison: |a - b| <= atol + rtol * |b| */
            double diff = fabs(val_a - val_b);
            double threshold = atol + rtol * fabs(val_b);
            out[i] = (diff <= threshold) ? 1 : 0;
        } else {
            /* b is inf but a is not exactly equal */
            out[i] = 0;
        }

        /* Increment indices */
        for (int d = bc_ndim - 1; d >= 0; d--) {
            indices[d]++;
            if (indices[d] < bc_shape[d]) break;
            indices[d] = 0;
        }
    }

    return result;
}

EXPORT int ndarray_allclose(NDArray* a, NDArray* b, double rtol, double atol, int equal_nan)
{
    NDArray* close = ndarray_isclose(a, b, rtol, atol, equal_nan);
    if (!close) return 0;

    NDArray* all_result = ndarray_all(close);
    ndarray_free(close);

    if (!all_result) return 0;

    int result = (ndarray_get_flat(all_result, 0) != 0.0) ? 1 : 0;
    ndarray_free(all_result);

    return result;
}

EXPORT int ndarray_array_equal(NDArray* a1, NDArray* a2, int equal_nan)
{
    if (!a1 || !a2) return 0;

    /* Check shape equality */
    if (a1->ndim != a2->ndim) return 0;

    for (int i = 0; i < a1->ndim; i++) {
        if (a1->shape[i] != a2->shape[i]) return 0;
    }

    /* Check element equality */
    for (size_t i = 0; i < a1->size; i++) {
        double v1 = ndarray_get_flat(a1, i);
        double v2 = ndarray_get_flat(a2, i);

        int v1_nan = isnan(v1);
        int v2_nan = isnan(v2);

        if (v1_nan && v2_nan) {
            if (!equal_nan) return 0;
            /* else: NaN == NaN is OK when equal_nan is true */
        } else if (v1_nan || v2_nan) {
            return 0;
        } else if (v1 != v2) {
            return 0;
        }
    }

    return 1;
}

EXPORT int ndarray_array_equiv(NDArray* a1, NDArray* a2)
{
    if (!a1 || !a2) return 0;

    /* Check broadcast compatibility */
    int32_t bc_shape[NPY_MAXDIMS];
    int32_t bc_ndim;

    if (broadcast_shapes(a1->shape, a1->ndim, a2->shape, a2->ndim, bc_shape, &bc_ndim) != 0) {
        return 0;
    }

    /* Compute broadcast strides */
    int32_t a1_strides[NPY_MAXDIMS];
    int32_t a2_strides[NPY_MAXDIMS];

    if (broadcast_strides(a1, bc_shape, bc_ndim, a1_strides) != 0 ||
        broadcast_strides(a2, bc_shape, bc_ndim, a2_strides) != 0) {
        return 0;
    }

    size_t a1_elem = dtype_size(a1->dtype);
    size_t a2_elem = dtype_size(a2->dtype);

    int32_t indices[NPY_MAXDIMS] = {0};
    size_t total = 1;
    for (int i = 0; i < bc_ndim; i++) total *= (size_t)bc_shape[i];

    for (size_t i = 0; i < total; i++) {
        size_t o1 = 0, o2 = 0;

        for (int d = 0; d < bc_ndim; d++) {
            if (a1_strides[d] != 0) {
                o1 += (size_t)indices[d] * ((size_t)a1_strides[d] / a1_elem);
            }
            if (a2_strides[d] != 0) {
                o2 += (size_t)indices[d] * ((size_t)a2_strides[d] / a2_elem);
            }
        }

        double v1 = ndarray_get_flat(a1, o1);
        double v2 = ndarray_get_flat(a2, o2);

        if (v1 != v2) return 0;

        /* Increment indices */
        for (int d = bc_ndim - 1; d >= 0; d--) {
            indices[d]++;
            if (indices[d] < bc_shape[d]) break;
            indices[d] = 0;
        }
    }

    return 1;
}

File: packages/numwasm/src/ts/logic.ts
/**
 * NumJS Logic & Comparison Functions
 *
 * TypeScript wrappers for WASM logic and comparison operations.
 * Follows NumPy's logic and comparison API.
 *
 * Reference: numpy/_core/fromnumeric.py, numpy/_core/numeric.py
 */

import { NDArray } from "./_core/NDArray.js";
import { DType } from "./types.js";
import { absolute } from "./ufunc.js";

/* ============ Element-wise Predicates ============ */

/**
 * Test element-wise for finiteness (not infinity and not NaN).
 *
 * @param x - Input array
 * @returns Boolean array where True indicates finite value
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([1, Infinity, -Infinity, NaN]);
 * const result = await isfinite(arr);  // [true, false, false, false]
 * ```
 */
export async function isfinite(x: NDArray): Promise<NDArray> {
  const module = x._wasmModule;
  const resultPtr = module._ndarray_isfinite(x._wasmPtr);
  if (resultPtr === 0) {
    throw new Error("isfinite failed");
  }
  return NDArray._fromPtr(resultPtr, module);
}

/**
 * Test element-wise for positive or negative infinity.
 *
 * @param x - Input array
 * @returns Boolean array where True indicates infinity
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([1, Infinity, -Infinity, NaN]);
 * const result = await isinf(arr);  // [false, true, true, false]
 * ```
 */
export async function isinf(x: NDArray): Promise<NDArray> {
  const module = x._wasmModule;
  const resultPtr = module._ndarray_isinf(x._wasmPtr);
  if (resultPtr === 0) {
    throw new Error("isinf failed");
  }
  return NDArray._fromPtr(resultPtr, module);
}

/**
 * Test element-wise for NaN.
 *
 * @param x - Input array
 * @returns Boolean array where True indicates NaN
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([1, Infinity, NaN]);
 * const result = await isnan(arr);  // [false, false, true]
 * ```
 */
export async function isnan(x: NDArray): Promise<NDArray> {
  const module = x._wasmModule;
  const resultPtr = module._ndarray_isnan(x._wasmPtr);
  if (resultPtr === 0) {
    throw new Error("isnan failed");
  }
  return NDArray._fromPtr(resultPtr, module);
}

/**
 * Test element-wise for negative infinity.
 *
 * @param x - Input array
 * @returns Boolean array where True indicates negative infinity
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([1, Infinity, -Infinity, NaN]);
 * const result = await isneginf(arr);  // [false, false, true, false]
 * ```
 */
export async function isneginf(x: NDArray): Promise<NDArray> {
  const module = x._wasmModule;
  const resultPtr = module._ndarray_isneginf(x._wasmPtr);
  if (resultPtr === 0) {
    throw new Error("isneginf failed");
  }
  return NDArray._fromPtr(resultPtr, module);
}

/**
 * Test element-wise for positive infinity.
 *
 * @param x - Input array
 * @returns Boolean array where True indicates positive infinity
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([1, Infinity, -Infinity, NaN]);
 * const result = await isposinf(arr);  // [false, true, false, false]
 * ```
 */
export async function isposinf(x: NDArray): Promise<NDArray> {
  const module = x._wasmModule;
  const resultPtr = module._ndarray_isposinf(x._wasmPtr);
  if (resultPtr === 0) {
    throw new Error("isposinf failed");
  }
  return NDArray._fromPtr(resultPtr, module);
}

/**
 * Returns a bool array where True if input element is complex (imaginary part != 0).
 * For non-complex dtype arrays, returns all False.
 *
 * @param x - Input array
 * @returns Boolean array
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([1+2j, 3+0j], { dtype: DType.Complex128 });
 * const result = await iscomplex(arr);  // [true, false]
 * ```
 */
export async function iscomplex(x: NDArray): Promise<NDArray> {
  const module = x._wasmModule;
  const resultPtr = module._ndarray_iscomplex_elem(x._wasmPtr);
  if (resultPtr === 0) {
    throw new Error("iscomplex failed");
  }
  return NDArray._fromPtr(resultPtr, module);
}

/**
 * Returns a bool array where True if input element is real (imaginary part == 0).
 * For non-complex dtype arrays, returns all True.
 *
 * @param x - Input array
 * @returns Boolean array
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([1+2j, 3+0j], { dtype: DType.Complex128 });
 * const result = await isreal(arr);  // [false, true]
 * ```
 */
export async function isreal(x: NDArray): Promise<NDArray> {
  const module = x._wasmModule;
  const resultPtr = module._ndarray_isreal_elem(x._wasmPtr);
  if (resultPtr === 0) {
    throw new Error("isreal failed");
  }
  return NDArray._fromPtr(resultPtr, module);
}

/* ============ Type Checking (TypeScript only) ============ */

/**
 * Check if the array has a complex dtype.
 *
 * @param x - Input array
 * @returns True if dtype is Complex64 or Complex128
 *
 * @example
 * ```typescript
 * const arr = await NDArray.zeros([2, 2], { dtype: DType.Complex128 });
 * iscomplexobj(arr);  // true
 * ```
 */
export function iscomplexobj(x: NDArray): boolean {
  return x.dtype === DType.Complex64 || x.dtype === DType.Complex128;
}

/**
 * Check if the array has a non-complex dtype.
 *
 * @param x - Input array
 * @returns True if dtype is not complex
 *
 * @example
 * ```typescript
 * const arr = await NDArray.zeros([2, 2]);
 * isrealobj(arr);  // true
 * ```
 */
export function isrealobj(x: NDArray): boolean {
  return !iscomplexobj(x);
}

/**
 * Check if the array is Fortran contiguous but not C contiguous.
 *
 * @param a - Input array
 * @returns True if F-contiguous and not C-contiguous
 *
 * @example
 * ```typescript
 * const arr = await NDArray.zeros([3, 4]);
 * const fortran = arr.asfortranarray();
 * isfortran(fortran);  // true
 * ```
 */
export function isfortran(a: NDArray): boolean {
  const flags = a.flags;
  return flags.f_contiguous && !flags.c_contiguous;
}

/**
 * Check if an element is a scalar type.
 *
 * @param element - Value to check
 * @returns True if element is number, boolean, or bigint
 *
 * @example
 * ```typescript
 * isscalar(5);        // true
 * isscalar(3.14);     // true
 * isscalar(true);     // true
 * isscalar([1, 2]);   // false
 * ```
 */
export function isscalar(element: unknown): boolean {
  return (
    typeof element === "number" ||
    typeof element === "boolean" ||
    typeof element === "bigint"
  );
}

/* ============ Truth Testing (Reductions) ============ */

/**
 * Test whether all array elements along a given axis evaluate to True.
 *
 * @param a - Input array
 * @param axis - Axis along which to test (undefined for entire array)
 * @param keepdims - If true, keep reduced axis as size-1 dimension
 * @returns Boolean result (scalar or array)
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([[true, true], [true, false]]);
 * await all(arr);           // false
 * await all(arr, 0);        // [true, false]
 * await all(arr, 1);        // [true, false]
 * ```
 */
export async function all(
  a: NDArray,
  axis?: number,
  keepdims: boolean = false,
): Promise<NDArray | boolean> {
  const module = a._wasmModule;

  let resultPtr: number;

  if (axis === undefined) {
    resultPtr = module._ndarray_all(a._wasmPtr);
  } else {
    resultPtr = module._ndarray_all_axis(a._wasmPtr, axis, keepdims ? 1 : 0);
  }

  if (resultPtr === 0) {
    throw new Error("all failed");
  }

  const result = NDArray._fromPtr(resultPtr, module);

  // For scalar result without keepdims, return boolean
  if (axis === undefined || result.ndim === 0) {
    const val = result.item() !== 0;
    result.dispose();
    return val;
  }

  return result;
}

/**
 * Test whether any array element along a given axis evaluates to True.
 *
 * @param a - Input array
 * @param axis - Axis along which to test (undefined for entire array)
 * @param keepdims - If true, keep reduced axis as size-1 dimension
 * @returns Boolean result (scalar or array)
 *
 * @example
 * ```typescript
 * const arr = await NDArray.fromArray([[false, false], [false, true]]);
 * await any(arr);           // true
 * await any(arr, 0);        // [false, true]
 * await any(arr, 1);        // [false, true]
 * ```
 */
export async function any(
  a: NDArray,
  axis?: number,
  keepdims: boolean = false,
): Promise<NDArray | boolean> {
  const module = a._wasmModule;

  let resultPtr: number;

  if (axis === undefined) {
    resultPtr = module._ndarray_any(a._wasmPtr);
  } else {
    resultPtr = module._ndarray_any_axis(a._wasmPtr, axis, keepdims ? 1 : 0);
  }

  if (resultPtr === 0) {
    throw new Error("any failed");
  }

  const result = NDArray._fromPtr(resultPtr, module);

  // For scalar result without keepdims, return boolean
  if (axis === undefined || result.ndim === 0) {
    const val = result.item() !== 0;
    result.dispose();
    return val;
  }

  return result;
}

/* ============ Comparison Functions ============ */

/**
 * Returns a boolean array where two arrays are element-wise equal within tolerance.
 *
 * Formula: |a - b| <= (atol + rtol * |b|)
 *
 * @param a - First array
 * @param b - Second array (reference)
 * @param rtol - Relative tolerance (default 1e-5)
 * @param atol - Absolute tolerance (default 1e-8)
 * @param equal_nan - Whether to compare NaN's as equal (default false)
 * @returns Boolean array of element-wise close status
 *
 * @example
 * ```typescript
 * const a = await NDArray.fromArray([1e10, 1e-7]);
 * const b = await NDArray.fromArray([1.00001e10, 1e-8]);
 * const result = await isclose(a, b);  // [true, false]
 * ```
 */
export async function isclose(
  a: NDArray,
  b: NDArray,
  rtol: number = 1e-5,
  atol: number = 1e-8,
  equal_nan: boolean = false,
): Promise<NDArray> {
  const module = a._wasmModule;

  const resultPtr = module._ndarray_isclose(
    a._wasmPtr,
    b._wasmPtr,
    rtol,
    atol,
    equal_nan ? 1 : 0,
  );

  if (resultPtr === 0) {
    throw new Error("isclose failed: shapes may not be broadcast-compatible");
  }

  return NDArray._fromPtr(resultPtr, module);
}

/**
 * Returns True if two arrays are element-wise equal within a tolerance.
 *
 * @param a - First array
 * @param b - Second array
 * @param rtol - Relative tolerance (default 1e-5)
 * @param atol - Absolute tolerance (default 1e-8)
 * @param equal_nan - Whether to compare NaN's as equal (default false)
 * @returns True if all elements are close
 *
 * @example
 * ```typescript
 * const a = await NDArray.fromArray([1e10, 1e-8]);
 * const b = await NDArray.fromArray([1.00001e10, 1e-9]);
 * await allclose(a, b);  // true
 * ```
 */
export async function allclose(
  a: NDArray,
  b: NDArray,
  rtol: number = 1e-5,
  atol: number = 1e-8,
  equal_nan: boolean = false,
): Promise<boolean> {
  const module = a._wasmModule;

  const result = module._ndarray_allclose(
    a._wasmPtr,
    b._wasmPtr,
    rtol,
    atol,
    equal_nan ? 1 : 0,
  );

  return result !== 0;
}

/**
 * True if two arrays have the same shape and elements, False otherwise.
 *
 * @param a1 - First array
 * @param a2 - Second array
 * @param equal_nan - Whether to compare NaN's as equal (default false)
 * @returns True if arrays are equal
 *
 * @example
 * ```typescript
 * const a = await NDArray.fromArray([1, 2, 3]);
 * const b = await NDArray.fromArray([1, 2, 3]);
 * await array_equal(a, b);  // true
 * ```
 */
export async function array_equal(
  a1: NDArray,
  a2: NDArray,
  equal_nan: boolean = false,
): Promise<boolean> {
  const module = a1._wasmModule;

  const result = module._ndarray_array_equal(
    a1._wasmPtr,
    a2._wasmPtr,
    equal_nan ? 1 : 0,
  );

  return result !== 0;
}

/**
 * Returns True if input arrays are shape consistent and all elements equal.
 *
 * Shape consistent means they are either the same shape, or one input array
 * can be broadcast to create the same shape as the other one.
 *
 * @param a1 - First array
 * @param a2 - Second array
 * @returns True if arrays are equivalent
 *
 * @example
 * ```typescript
 * const a = await NDArray.fromArray([1, 2]);
 * const b = await NDArray.fromArray([[1, 2], [1, 2]]);
 * await array_equiv(a, b);  // true
 * ```
 */
export async function array_equiv(a1: NDArray, a2: NDArray): Promise<boolean> {
  const module = a1._wasmModule;

  const result = module._ndarray_array_equiv(a1._wasmPtr, a2._wasmPtr);

  return result !== 0;
}

/**
 * If input is complex with all imaginary parts close to zero, return real parts.
 *
 * "Close to zero" is defined as `tol` times the machine epsilon of the type.
 *
 * @param a - Input array
 * @param tol - Tolerance in machine epsilons for complex128, or absolute
 *              tolerance if tol < 1. Default is 100.
 * @returns Real array if imaginary parts are negligible, otherwise the input
 *
 * @example
 * ```typescript
 * // Imaginary parts within tolerance - returns real
 * const c1 = await NDArray.fromArray([1+1e-16j, 2+1e-16j], { dtype: DType.Complex128 });
 * const r1 = await real_if_close(c1);  // [1, 2] (Float64)
 *
 * // Imaginary parts too large - returns complex unchanged
 * const c2 = await NDArray.fromArray([1+1j, 2+2j], { dtype: DType.Complex128 });
 * const r2 = await real_if_close(c2);  // [1+1j, 2+2j] (Complex128)
 * ```
 */
export async function real_if_close(
  a: NDArray,
  tol: number = 100,
): Promise<NDArray> {
  // For non-complex arrays, return as-is
  if (!iscomplexobj(a)) {
    return a;
  }

  // Get machine epsilon for this type
  // Float32 epsilon: ~1.19e-7, Float64 epsilon: ~2.22e-16
  const eps =
    a.dtype === DType.Complex64 ? 1.1920929e-7 : 2.220446049250313e-16;
  const tolerance = tol > 1 ? eps * tol : tol;

  // Check if all imaginary parts are below tolerance
  const imagPart = a.imag;
  const absImag = absolute(imagPart);

  // Find maximum absolute imaginary value
  const module = a._wasmModule;
  const size = absImag.size;
  let maxImag = 0;

  for (let i = 0; i < size; i++) {
    const val = module._ndarray_get_flat(absImag._wasmPtr, i);
    if (val > maxImag) {
      maxImag = val;
    }
  }

  if (maxImag < tolerance) {
    // All imaginary parts are negligible, return real parts
    return a.real;
  }

  // Imaginary parts are significant, return as-is
  return a;
}
